sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -57,26 +90,135 @@ case class GpuNot(child: Expression) extends CudfUnaryExpression
+abstract class CudfBinaryPredicateWithSideEffect extends CudfBinaryOperator
+  with Predicate with GpuCVPredicateHelper {
+  
   override def inputType: AbstractDataType = BooleanType
 
+  def columnarEvalWithSideEffects(batch: ColumnarBatch): Any
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val rightExpr = right.asInstanceOf[GpuExpression]
+
+    if (rightExpr.hasSideEffects) {
+      columnarEvalWithSideEffects(batch)
+    } else {
+      super.columnarEval(batch)
+    }
+  }
+}
+
+case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryPredicateWithSideEffect {
+  
   override def symbol: String = "&&"
 
   override def sqlOperator: String = "AND"
 
   override def binaryOp: BinaryOp = BinaryOp.NULL_LOGICAL_AND
   override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.NULL_LOGICAL_AND)
+
+
+  /**
+   * When computing logical expressions on the CPU, the true and false
+   * expressions are evaluated lazily, meaning that the RHS expression
+   * of logical-AND is not evaluated when LHS is False. For logical-OR,
+   * RHS is not evaluated when LHS is True.
+   * This is important in the case where the expressions can have
+   * side-effects, such as throwing exceptions for invalid inputs.
+   *
+   * This method performs lazy evaluation on the GPU by first filtering
+   * the input batch a  where the LHS predicate is True.
+   * The RHS predicate is evaluated against these batches and then the
+   * results are combined back into a single batch using the gather
+   * algorithm.
+   */
+  override def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val leftExpr = left.asInstanceOf[GpuExpression]
+    val rightExpr = right.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(GpuExpressionsUtils.columnarEvalToColumn(leftExpr, batch)) { lhsBool =>
+        if (isAllFalse(lhsBool)) {
+          withResource(GpuScalar.from(false, dataType)) { falseScalar =>
+            GpuColumnVector.from(falseScalar, lhsBool.getRowCount.toInt, BooleanType)
+          }
+        } else {
+          // replace nulll values
+          val lhsNoNulls = withResource(Scalar.fromBool(true)) { trueScalar =>
+            lhsBool.getBase.replaceNulls(trueScalar)
+          }
+          withResource(lhsNoNulls) { _ =>
+            val rEval = withResource(filterBatch(tbl, lhsNoNulls, colTypes)) { leftTrueBatch =>
+              rightExpr.columnarEval(leftTrueBatch)
+            }
+            val finalRet = withResourceIfAllowed(rEval) { _ =>
+              (rEval) match {
+                case (f: GpuColumnVector) =>
+                  withResource(gather(lhsNoNulls, f)) { combinedVector =>
+                    doColumnar(lhsBool, GpuColumnVector.from(combinedVector, dataType))
+                  }
+                case (f: GpuScalar) =>
+                  doColumnar(lhsBool, f)
```

Comment:
This path is not covered by the current tests. Would we ever hit this path? I don't think that scalar expressions can have side-effects and we would only call this method if the RHS has side-effects.

Reply:
Wow that is an interesting situation. The only place right now where we can return a Scalar from processing when the input is not a foldable constant is some corner cases with Coalesce, and Coalesce can totally have side effects associated with it for down stream tasks that are a part of it.  Perhaps we should try to write a test for this just to be sure...

Reply:
I've been trying to write a test that hits the scalar path and have failed to come up with anything so far. It looks like Coalesce will only return a scalar if the first non-null parameter is a scalar, but Spark will optimize the Coalesce out in that case. I am back to wondering if the scalar path is even possible, but maybe I am missing the edge case here.

Reply:
I don't think Spark had an optimization in older versions, so you might try that.  Also this is enough of a corner case that I am fine if it as is.